src/pymor/algorithms/timestepping.py

```diff
@@ -10,44 +10,363 @@
-:func:`implicit_euler` to provide this interface.
+time-stepping are provided, based on the :class:`ExplicitEulerTimeStepper` and
+:class:`ImplicitEulerTimeStepper` classes. These derive from :class:`TimeStepper`,
+which defines a common interface that has to be fulfilled by the time-steppers used
+by |InstationaryModel|.
 """
 
+from numbers import Number
+import numpy as np
+
 from pymor.core.base import ImmutableObject, abstractmethod
+from pymor.operators.constructions import IdentityOperator, VectorArrayOperator, ZeroOperator
 from pymor.operators.interface import Operator
+from pymor.parameters.base import Mu
+from pymor.tools import floatcmp
 from pymor.vectorarrays.interface import VectorArray
 
 
 class TimeStepper(ImmutableObject):
     """Interface for time-stepping algorithms.
 
-    Algorithms implementing this interface solve time-dependent problems
+    Algorithms implementing this interface solve time-dependent initial value problems
     of the form ::
 
-        M * d_t u + A(u, mu, t) = F(mu, t).
+        M * d_t u + A(u, mu, t) = F(mu, t),
+                     u(mu, t_0) = u_0(mu).
+
+    Time-steppers used by |InstationaryModel| have to fulfill this interface. Time evolution can either be performed
+    by calling :meth:`solve`, or by manually calling :meth:`bootstrap` and :meth:`step`.
 
-    Time-steppers used by |InstationaryModel| have to fulfill
-    this interface.
+    Note that :meth:`bootstrap` and :meth:`step` are designed to support a minimal implementation of :meth:`solve`,
+    as the outer loop in :meth:`solve` may arise manually in several places.
+
+    Parameters
+    ----------
+    initial_time
+        The time at which to begin time-stepping.
+    end_time
+        The time until which to perform time-stepping.
+    num_values
+        The number of returned vectors of the solution trajectory. If `None`, each intermediate vector that is
+        calculated is returned. Else an interpolation of the calculated vectors on an equidistant temporal grid is
+        returned, using an appropriate interpolation of the respective time stepper.
+
+    Attributes
+    ----------
+    steps
+        The number of previous steps required to determine the next. I.e. steps == 1 for a single-step method and
+        steps > 1 for a multi-step method.
     """
 
+    initial_time = None
+    end_time = None
+    steps = None
+    implicit = False
+    num_values = None
+
+    def __init__(self, initial_time, end_time, num_values=None):
+        assert isinstance(initial_time, Number)
+        assert isinstance(end_time, Number)
+        assert end_time > initial_time
+        assert not num_values or (isinstance(num_values, Number) and num_values > 1)
+        if num_values:
+            self._save_every = (end_time - initial_time) / (num_values - 1)
+        self.__auto_init(locals())
+
+    def bootstrap(self, initial_data, operator, rhs=None, mass=None, mu=None, reserve=False):
```

Comment:
I believe I have written this before. I would really prefer to give `solve` and optional `iter` argument. If `True`, `solve` would return an iterator that stores all the data returned by `bootstrap`.

Reply:
I am totally in favor of this, I just did not do it yet.